### ISSUE TITLE
DIFFs for LLMs

### ISSUE THREAD

**Issue by dazzaji** [dazzaji/pubs#1]
# LLM Diffs

From Campbell:  

"i ended up doing something similar to Aider diffs with the <<<<<<< SEARCH open tag and then —— to separate and then the replace"

"this was because there is a unix tool that operates similarly - so i thought it seemed like a good idea"

## o1 Pro

Below are instructions for using diff formats that LLMs can understand. We’ll start by showing how to use Aider’s approach, then show how to do something similar to your friend’s approach. The goal is to let the LLM know exactly which part of the file you want changed and what the change should be. You will then prompt the LLM with this format and ask it to provide the relevant output.

Using Aider’s Approach

Basic Idea:
You present the LLM with a section of code, mark the exact spot you want changed using a “SEARCH” block, and then show what you want replaced with a “REPLACE” block. The LLM understands that it should transform the “SEARCH” code into the “REPLACE” code. After providing these instructions, you ask the LLM for the updated code snippet.

Step-by-Step Instructions:
	1.	Identify the file and code snippet you want to modify.
	2.	Write a “SEARCH/REPLACE” style diff using <<<<<<< SEARCH and >>>>>>> REPLACE.
	3.	Insert the code you want to find under SEARCH and the code you want to replace it with under REPLACE.
	4.	Give this entire snippet to the LLM and instruct it to apply the changes.

Example Using Aider’s Approach:
Suppose you have a Python file app.py and you currently have:

from flask import Flask

and you want to import math as well. You could present the LLM with:

app.py

<<<<<<< SEARCH
from flask import Flask
=======
import math
from flask import Flask
>>>>>>> REPLACE

What to Tell the LLM:
	•	In your prompt, say:
“Here is a diff in the Aider style. Please apply the changes indicated by the SEARCH/REPLACE blocks and return the updated app.py code.”

Result You’ll Get:
The LLM should return the updated code:

import math
from flask import Flask

Using Your Friend’s Approach

Your friend mentioned doing something similar but using a slightly different delimiter style. Instead of the <<<<<<< SEARCH and >>>>>>> REPLACE tags, they might use something like <<<<<< SEARCH and then a separator line ------ before the replacement content, and finally a tag like REPLACE at the end. This still signals the LLM where to find the original code and what to replace it with, but with slightly different formatting.

Step-by-Step Instructions:
	1.	Identify the code snippet you want to change.
	2.	Use a <<<<<< SEARCH line to indicate where the original code starts.
	3.	After showing the original code, insert a line of ------ to separate the original from the replacement.
	4.	Then write the replacement code.
	5.	Close it off with something like REPLACE.

Example Using Your Friend’s Approach:
Let’s say you have the same app.py file and you want to make the same change as above:

app.py

<<<<<< SEARCH
from flask import Flask
------
import math
from flask import Flask
REPLACE

What to Tell the LLM:
	•	In your prompt, say:
“Here is a diff using my friend’s approach. The original code is under SEARCH, then after the ‘——’ line is what I’d like to replace it with. Please apply these changes and return the updated app.py.”

Result You’ll Get:
Again, the LLM should return:

import math
from flask import Flask

General Instructions for Prompting the LLM
	•	Always include the filename and show the LLM the exact code snippet you’re targeting.
	•	Use consistent markers (SEARCH and REPLACE) so the LLM can easily parse the instructions.
	•	Keep the format simple and minimal—just enough for the LLM to understand the original and desired code.
	•	After providing the diff, explicitly ask the LLM to return the updated code snippet.

Example Prompt:

“Below is a code diff showing what I want changed. Use the format provided and produce the updated code:

app.py

<<<<<<< SEARCH
from flask import Flask
=======
import math
from flask import Flask
>>>>>>> REPLACE

Please return the updated code.”

You can swap out the markers for your friend’s style, or any similar style, as long as it’s clear to the LLM what should be replaced and with what.

By following these instructions and examples, you can give LLMs diffs in a format they understand. You can then instruct the LLM to apply the changes and return the updated code.

——

## 4o

Here’s how you can adapt both Aider’s approach and your friend’s approach for handling diffs with large language models (LLMs). This will allow you to structure inputs for clear understanding and output generation by the LLM.

Step-by-Step Instructions

1. Define Your Goals for Diffing with LLMs

You want the LLM to:
	•	Understand the original vs. modified content clearly.
	•	Suggest or apply relevant changes based on diffs.
	•	Return clean, token-efficient diffs in its output.

2. Approach 1: Using the Aider Diff Format

Aider uses a clear SEARCH/REPLACE diff format, leveraging a familiar structure for both humans and LLMs. Here’s how you can apply it:

Input Instructions for the LLM
	1.	Provide the filename or context.
	2.	Mark the sections to change with:
	•	<<<<<<< SEARCH to show the existing code/content.
	•	======= as a separator.
	•	>>>>>>> REPLACE to show the desired updated content.

Example Prompt Using Aider’s Approach

Let’s edit a Python function to improve efficiency:

Input to LLM:

Filename: optimize_calculations.py

<<<<< SEARCH
for i in range(len(data)):
    result.append(data[i] * 2)
=======
result = [x * 2 for x in data]
>>>>> REPLACE

Please apply this change and explain why it improves efficiency.

Expected LLM Output:

Change applied to optimize_calculations.py:
The loop was replaced with a list comprehension for improved readability and performance.

Updated code:
result = [x * 2 for x in data]

Why This Works
	•	The format mimics tools like git merge conflict markers, which LLMs trained on code are familiar with.
	•	It minimizes token usage by showing only the changed sections instead of the entire file.

3. Approach 2: Your Friend’s Diff Format

Your friend’s approach uses a simpler structure inspired by Unix tools, using <<<<< SEARCH, a separator like ----, and the replacement content.

Input Instructions for the LLM
	1.	Mark the existing content with <<<<< SEARCH.
	2.	Use ---- as a separator.
	3.	Provide the replacement content after the separator.

Example Prompt Using Your Friend’s Approach

Continuing with the same Python function:

Input to LLM:

Filename: optimize_calculations.py

<<<<< SEARCH
for i in range(len(data)):
    result.append(data[i] * 2)
----
result = [x * 2 for x in data]

Please make this edit and explain the benefit.

Expected LLM Output:

Edit applied to optimize_calculations.py:
The loop was converted to a list comprehension, reducing the code's verbosity and improving execution speed.

Updated code:
result = [x * 2 for x in data]

Why This Works
	•	Simplicity: The use of ---- as a separator is straightforward for both humans and LLMs.
	•	Compatibility: This format aligns with Unix-inspired tools, which may be part of the LLM’s training data.

4. Comparing the Two Approaches

Feature	Aider’s Approach	Friend’s Approach
Clarity	High (uses Git-like, familiar markers)	Moderate (simple but less explicit markers)
Token Efficiency	High (compact, precise change representation)	High (minimal overhead in formatting)
Ease of Implementation	Moderate (requires precise formatting)	High (straightforward structure)
LLM Familiarity	High (resembles Git conflict resolution)	High (Unix-style simplicity)

5. General Guidelines for Using Diffs with LLMs
	1.	Focus on Token Efficiency: Show only relevant parts of the code or document. Avoid including unnecessary context.
	2.	Structure for Readability: Use consistent, clear markers (like SEARCH and REPLACE or ----) to make diffs easy for both humans and LLMs to parse.
	3.	Provide Explicit Instructions: Clearly state what you want the LLM to do with the diff (e.g., apply changes, explain differences, suggest improvements).

Final Example: Combining Aider and Your Friend’s Format

If you want flexibility, consider using both approaches together. Here’s how:

Input to LLM:

Filename: optimize_calculations.py

Aider-style diff:
<<<<< SEARCH
for i in range(len(data)):
    result.append(data[i] * 2)
=======
result = [x * 2 for x in data]
>>>>> REPLACE

Friend-style diff:
<<<<< SEARCH
for i in range(len(data)):
    result.append(data[i] * 2)
----
result = [x * 2 for x in data]

Please evaluate both formats and suggest which is better for this scenario.

Expected LLM Output:

Both formats are effective. The Aider-style diff is more explicit and may be better for complex changes involving multiple files. The Friend-style diff is simpler and works well for single-file, small edits.

Recommended change:
result = [x * 2 for x in data]

This ensures you and the LLM have flexible methods for different scenarios.

----------------

# SOURCE INFO: Perplexity on Aider and Cursor Approaches to DIFFs

Cursor and Aider have indeed pioneered innovative approaches to using LLM-friendly diff formats to improve code editing accuracy and efficiency. Both tools have developed unique input/output methods to enhance the quality of LLM-generated code changes. Let's explore how they handle diffs for LLMs:

## Aider's Approach

Aider uses a specialized diff format called "SEARCH/REPLACE" to communicate code changes to LLMs[1]. This format is designed to be efficient and reduce errors in code editing:

- It specifies file edits as a series of search/replace blocks[1].
- The format is similar to git merge conflict resolution markings[1].
- It only requires the LLM to return parts of the file that have changes, reducing token usage[1].

Example of Aider's diff format:

```
mathweb/flask/app.py
```
<<<<<<< SEARCH
from flask import Flask
=======
import math
from flask import Flask
>>>>>>> REPLACE
```

Aider has also introduced other techniques to improve code editing:

- Tree-sitter integration and LLM-friendly source map as global context[2].
- Multi-file edits[2].
- Running lint checks and unit tests, then feeding results back to the LLM for autocorrection[2].

## Cursor's Approach

Cursor has developed its own methods for handling code edits:

- They perform "magic" in the back-end to ensure the best pluggability of generated code[2].
- Cursor uses a proprietary model for handling diffs, as current GPT-4 doesn't handle it reliably yet[9].
- For files under 400 lines, Cursor found that fully rewriting the full file outperforms Aider-like diffs[10].

Cursor has also implemented other features to enhance code editing:

- Conceptual thinking about autocomplete and multi-line edits[2].
- Fine-tuning their own models for features like Copilot++[2].
- Experimenting with shadow workspaces to give LLMs a developer's perspective[2].

## Comparison and Evolution

Both tools have evolved their approaches over time:

1. **Efficiency**: Aider's diff format is designed to be token-efficient, while Cursor's approach of full file rewriting for smaller files may prioritize accuracy over token usage[1][10].

2. **Context Management**: Both tools have developed methods to manage context effectively. Aider uses tree-sitter and source maps, while Cursor experiments with shadow workspaces[2].

3. **Multi-file Edits**: Both tools have implemented capabilities for handling changes across multiple files[2].

4. **Feedback Loops**: Aider incorporates lint checks and unit tests into its workflow, while Cursor uses shadow workspaces to simulate a developer's environment[2].

5. **Model Specialization**: Cursor has developed proprietary models specifically for handling code edits, while Aider focuses on optimizing prompts and formats for existing LLMs[2][9].

In conclusion, both Cursor and Aider have developed sophisticated methods for handling diffs and code edits with LLMs. Their approaches differ in some aspects, but both aim to improve the accuracy and efficiency of AI-assisted coding. The choice between them may depend on specific use cases and preferences for token efficiency versus full file rewriting for smaller files.

Sources
[1] Edit formats - Aider https://aider.chat/docs/more/edit-formats.html
[2] Continue.dev: The Swiss Army Knife That Sometimes Fails to Cut https://dev.to/maximsaplin/continuedev-the-swiss-army-knife-that-sometimes-fails-to-cut-4gg3
[3] Code output in git diff instead of full code - Cursor - Community Forum https://forum.cursor.com/t/code-output-in-git-diff-instead-of-full-code/6556
[4] CursorCore: Assist Programming through Aligning Anything - arXiv https://arxiv.org/html/2410.07002v1
[5] LLM codeblock diff for merging algorithm : r/LocalLLaMA - Reddit https://www.reddit.com/r/LocalLLaMA/comments/1f7y47i/llm_codeblock_diff_for_merging_algorithm/
[6] Help me understand why I'd use anything beyond Cursor? - Reddit https://www.reddit.com/r/ChatGPTCoding/comments/1h2nf7q/help_me_understand_why_id_use_anything_beyond/
[7] NEW Aider Architect & Editor Updates Are INSANE!  (Beast Cursor ... https://www.youtube.com/watch?v=8jD8dAXq8jE
[8] How I use LLMs | Karan Sharma https://mrkaran.dev/posts/using-llm/
[9] Has Cursor gotten "dumb" - Discussion https://forum.cursor.com/t/has-cursor-gotten-dumb/2884
[10] Explore cursor.sh's 'fast apply' techniques ("fully rewriting the full file ... https://github.com/Aider-AI/aider/issues/625

----

## Deeper Dive into Cursor Approach (or whatever we can glean about it at least)

Cursor has taken a different approach to handling code edits compared to Aider's diff-based method. Here's how Cursor handles code changes for LLMs:

### Full-File Rewrites

Cursor opted for full-file rewrites instead of using diffs for several reasons:

1. **More Tokens for Processing:** By rewriting entire files, the model has more tokens to work with, allowing for more comprehensive thinking and problem-solving[1].

2. **Training Data Alignment:** Models are likely exposed to more full-file code examples during training than diff-formatted ones, making this approach more natural for them[1].

3. **Line Number Challenges:** Diff formats often struggle with line numbers, especially when tokenizers treat them as single tokens. Full-file rewrites eliminate this issue[1].

## Speculative Edits

Cursor implemented a technique called "speculative edits" to improve the speed of full-file rewrites:

1. **Faster Processing:** This method allows for speeds up to 9x faster than traditional approaches[1].

2. **Chunk-Based Predictions:** Instead of generating tokens one by one, Cursor's algorithm predicts larger chunks of the file at once[1].

3. **Deterministic Algorithm:** Cursor uses a deterministic algorithm to speculate on future tokens, optimizing for their fast-apply model[1].

### File Size Considerations

- For files under 400 lines, Cursor found that fully rewriting the entire file outperforms diff-based approaches like Aider's[10].
- This approach prioritizes accuracy over token efficiency for smaller files[1].

### Proprietary Model

Cursor has developed its own proprietary model for handling code edits, as they found current GPT-4 doesn't handle diffs reliably[9].

### Visual Representation

Cursor uses a visual diff interface to show suggested changes:

1. **Color-Coded Changes:** Modifications are displayed with red and green highlighting in the chat window[4].
2. **Interactive Application:** Users can apply and review diffs, with the ability to accept changes directly in the interface[4].

### Multiple Diff Interfaces

Cursor has optimized different diff interfaces for various scenarios:

1. **Autocomplete:** A specialized diff interface for quick suggestions during typing[4].
2. **Larger Code Blocks:** A separate interface for reviewing more substantial code changes[4].
3. **Multi-File Edits:** An optimized diff system for handling changes across multiple files[4].

By focusing on full-file rewrites and developing specialized models and interfaces, Cursor aims to provide a more accurate and efficient code editing experience, especially for smaller files and quick suggestions.

Sources
[1] How Cursor AI Implemented Instant Apply: File Editing at 1000 ... https://blog.getbind.co/2024/10/02/how-cursor-ai-implemented-instant-apply-file-editing-at-1000-tokens-per-second/
[2] Code output in git diff instead of full code - Cursor - Community Forum https://forum.cursor.com/t/code-output-in-git-diff-instead-of-full-code/6556
[3] LLM codeblock diff for merging algorithm : r/LocalLLaMA - Reddit https://www.reddit.com/r/LocalLLaMA/comments/1f7y47i/llm_codeblock_diff_for_merging_algorithm/
[4] Code diff: Merging code is hard | Cursor Team and Lex Fridman https://www.youtube.com/watch?v=6MKFOFOhEak
[5] How do inline suggestions with specific character deletions and ... https://forum.cursor.com/t/how-do-inline-suggestions-with-specific-character-deletions-and-additions-work-in-cursor-editor/26796
[6] A Month with Cursor and Claude-Dev: My Thoughts https://jasonroell.com/2024/10/10/a-month-with-cursor-and-claude-dev-my-thoughts/
[7] Ask HN: Best practices for AI-generated code diff outputs? - Reddit https://www.reddit.com/r/ChatGPTCoding/comments/1f4oaj3/ask_hn_best_practices_for_aigenerated_code_diff/
[8] Thoughts on coding with LLMs https://www.vipshek.com/blog/cursor
[9] Cursor AI vs Copilot: A Detailed Analysis - Codoid https://codoid.com/ai/cursorai-vs-copilot-a-detailed-analysis/
[10] How can I apply AI-generated changes only to specific code lines I ... https://forum.cursor.com/t/how-can-i-apply-ai-generated-changes-only-to-specific-code-lines-ive-selected/20894
